28 add arm support

nuvious opened the pull request:
Added support to build specifically amd64, armv6, armv7, and arm64v8 images.

## CICD tested using my own dockerhub

[Build logs in github](https://github.com/nuvious/proxpi/actions/runs/8303614932/job/22728030306)

![image](https://github.com/EpicWink/proxpi/assets/5287736/4be7cd0a-278b-40e9-b7f7-203b6afaaf00)
[https://hub.docker.com/r/nuvious/proxpi/tags](https://hub.docker.com/r/nuvious/proxpi/tags)

## Arm container tested on my Kubernetes cluster comprising of 11 raspberry pi 4's

![image](https://github.com/EpicWink/proxpi/assets/5287736/31358ec2-c5f3-4682-b4f6-500765e8db5a)

Issue: #26 (My branch is called issue-28 in error)

nuvious commented on Dockerfile:
```diff
@@ -2,7 +2,7 @@ FROM python:3.12-alpine
 RUN --mount=source=.,target=/root/src/proxpi,rw \
     uname -a && cat /etc/issue && apk --version && python --version && pip --version \
- && apk --no-cache add git \
+ && apk --no-cache add git build-base libxslt-dev libxml2-dev \
```
I had build failures related to the lxml library which expects gcc and these libraries to be available to build. They likely don't build arm wheels so lxml needs to build from source.

nuvious commented on tests/test-docker-image.sh:
```diff
@@ -2,7 +2,7 @@
 # Test Docker image
 #
-# Usage: test-docker-image.sh TAG
+# Usage: test-docker-image.sh REPO/IMAGE:TAG
```
Just made this modification in case folks in the future want to work on the CICD besides me. Will prevent this file from having to be edited when you have contributors.

nuvious commented:
Performing one more test on my cluster with a reduced image size. I forgot to purge un-needed build packages at the end which increased the image size significantly.

nuvious commented:
Did a manual build/test on my raspberry pi as well to confirm the docker image test works on arm; github can only do that test on amd64 architecture:

![image](https://github.com/EpicWink/proxpi/assets/5287736/46885f38-cacb-4811-bed1-ef87a8fd184c)

Also corrected usage of apk to purge build-base and the libraries I imported to reduce the image size appropriately:

![image](https://github.com/EpicWink/proxpi/assets/5287736/1374afb3-efb6-4a5d-996a-fb73ecddcc7e)

Also re-tested in kubernetes on a pi cluster:

![image](https://github.com/EpicWink/proxpi/assets/5287736/ea7cdc7e-7146-40bf-94c0-eb29ba294a29)

EpicWink commented:
Thanks for the PR. The issue is the same as with #24 and #27, which is that I can't test the ARM image on every release (I don't own an ARM device, I don't have the time to set up ARM emulation, and most importantly, [GitHub doesn't have ARM CI runners](https://github.blog/changelog/2023-10-30-accelerate-your-ci-cd-with-arm-based-hosted-runners-in-github-actions/)).

benedikt-bartscher commented:
@EpicWink I could set up repository mirroring to my companys gitlab instance and grant maintainer access for you if that helps (for free). We have many x86 and arm runners with a dedicated buildkit cluster.

Thanks for maintaining proxpi!

nuvious commented:
I have an idea that should help you out then. Will get something working on my fork and update this PR accordingly.

nuvious commented:
Did some more modifications that run the unit and integration tests in the containers via qemu so you can utilize gtihub x86 runners to do arm testing. I removed the armv6 and armv7 configurations since they were not a part of the original request and because they add significant build time due to emulation speeds. They could be added back under a different issue later potentially.

Also made the job definitions as modular as possible. This means the arm64 and amd64 containers are built and tested on independent runners in parallel but the runner that builds and pushes the final docker image layers has to re-build the process. I tried some tricks to share the docker cache between jobs but nothing was saving time that I tried. This could be modified to run everything under one runner but would sacrifice modularity since the docker commands would require the `--platform` flag to be added and copied multiple times (or a script written for it).

Given the build only takes about 5 minutes in total that seemed sufficient enough for the task at hand and the modularity makes adding other architectures down the line a lot easier. 

Completed job:
![image](https://github.com/EpicWink/proxpi/assets/5287736/14422d0e-a7c2-4d4b-93f2-669e3a75701e)
https://github.com/nuvious/proxpi/actions/runs/8333123237

Look in the build logs for `aarch64` for confirmation everything was running under aarch64/arm64. I added `uname -a` to the unit test execution to make this visible:
![image](https://github.com/EpicWink/proxpi/assets/5287736/d6402362-4d59-4a40-a6fc-5a618e381252)

Images:
![image](https://github.com/EpicWink/proxpi/assets/5287736/d0cb12d9-2216-4658-9569-ae856b56ad53)

Let me know if you have further questions, direction or guidance on this effort and thanks for your response!

nuvious commented:
P.S. Was able to remove my modifications to the Dockerfile since the lxml library has arm64 support and the scope no longer includes armv6 or armv7. Build log for that test under my 'dev' tag:

https://github.com/nuvious/proxpi/actions/runs/8333385815
